### ISSUE TITLE
Unhandled rejected promise

### ISSUE THREAD

**Issue by poti1** [briandfoy/mojo-promise-role-higherorder#1]
When using the module and there is an overall rejection, this warning shows up:

"Unhandled rejected promise at ..."

Use case:

```
#!/usr/bin/perl

use warnings;
use strict;
use feature 'say';
use Mojo::Promise;
use Mojo::Promise::Role::HigherOrder;

sub pass { say "OK" }
sub fail { say "NOK" }

sub normal_promise {
	my @promises = map{Mojo::Promise->new} 1..3;
	my $promise  = Mojo::Promise->new
		->any(@promises)
		->then(\&pass, \&fail);

	$_->reject for @promises;

	$promise->wait;
}

sub buggy_promise {
	my @promises = map{Mojo::Promise->new} 1..3;
	my $promise  = Mojo::Promise->with_roles('+Any')->new
		->any(@promises)
		->then(\&pass, \&fail);

	$_->reject for @promises;

	$promise->wait;
}

normal_promise();
buggy_promise();
```

**Comment by poti1** [comment]
Here is the **any** section from Mojo::Promise function _all:

```
    # "any"
    elsif ($type == 3) {
      $promises[$i]->then(
        sub { $all->resolve(@_); () },
        sub {
          $results->[$i] = [@_];
          $all->reject(@$results) if --$remaining <= 0;
          return ();
        }
      );
    }
```

**Comment by poti1** [comment]
I noticed that the return value from resolve/reject is used by the function **_then_cb**:

```
sub _then_cb {
  my ($new, $cb, $method, @results) = @_;

  return $new->$method(@results) unless defined $cb;

  my @res;
  return $new->reject($@) unless eval { @res = $cb->(@results); 1 };
  return $new->resolve(@res);
}
```

**Comment by poti1** [comment]
Simply updating the code references to return a false value fixes this bug:

```
sub any {
	my( $self, @promises ) = @_;
	my $any = $self->new;

	my $count = 0;
	$_->then(
		sub { $any->resolve( @_ ) },
		sub { $count++; $any->reject if $count == @promises; () }  # RETURN FALSE
		) foreach @promises;

	return @promises ? $any : $any->reject;
	}
```

**Comment by poti1** [comment]
Similar for **none**:

```
sub none {
	my( $self, @promises ) = @_;
	my $none = $self->new;

	my $count = 0;
	$_->then(
		sub { $none->reject( @_ ); () }, # RETURN FALSE
		sub { $count++; $none->resolve if $count == @promises; () } # RETURN FALSE
		) foreach @promises;

	return @promises ? $none : $none->resolve;
	}
```

**Comment by poti1** [comment]
Similar for **some**:

```
sub some {
	my( $self, $promises, $n ) = @_;
	my $some = $self->new;

	return $n == 0 ? $some->resolve : $some->reject if @$promises == 0;
	return $some->reject if $n > @$promises;

	my $remaining = @$promises;

	my( @resolved, @rejected );
	foreach my $p ( @$promises ) {
		$p->then(
			sub {
				$remaining--;
				push @resolved, [ @_ ];
				$some->resolve( @resolved ) if @resolved == $n;
				(); # RETURN FALSE
				},
			sub {
				# I keep trying to come up with a situation where
				# you'd reject $some with fewer rejections, but as
				# long as the sum of the pending and fulfilled is at
				# least the minimum count, there's always a chance.
				# The only way for that to not be true is for there
				# to be enough rejections to make pending small enough.
				# That's PROMISES - N + 1. Let's see if I can understand
				# that the next time I come through this. I still don't
				# believe it.
				$remaining--;
				push @rejected, [ @_ ];
				$some->reject( @rejected ) if @rejected > @$promises - $n;
				(); # RETURN FALSE
				},
			)
		}

	return $some;
	}
```

**Comment by briandfoy** [comment]
This is fixed in 1.005 (30b0c285d2fbce7305d18588adaa26f347a4b257). Thanks,
